bpo-47015: Update test_os from asyncore to asyncio

arhadthedev opened the pull request:
[PEP 594 – Removing dead batteries from the standard library](https://peps.python.org/pep-0594/) removes asyncore and asynchat in 3.12 with the following note:

> The `asyncore` module is also used in stdlib tests. The tests for `ftplib`, `logging`, `smptd`, `smtplib`, and `ssl` are partly based on asyncore. These tests must be updated to use asyncio or threading.

However, the citation misses the tests for `os`. So this PR ports `test_os.TestSendfile` as the only class that uses asyncore to create a dummy sendfile server and communicate with it.

To simplify the review I broke all changes into a few commits to get readable diffs:

1. Preparation:
    1. Change inheritance of the test from `TestCase` to `IsolatedAsyncioTestCase`
    1. Make `os.sendfile` async-friendly by moving it into a worker thread to not block an event loop of IsolatedAsyncioTestCase. As a result, `sendfile_wrapper` that called it is also made async
    1. Explicitly specify a global event loop policy because an implicit initialization
 by `IsolatedAsyncioTestCase` is counted as a test failure (this commit is actually the last)
1. Replace `SendfileTestServer` with `asyncio.start_server`. The client side is left mostly untouched because `os.sendfile` works directly with a low-level socket ignoring asyncio message loop anyway. "Mostly" because `connect` is delegated to an event loop to not block the server in the same thread
1. Cleanup:
    1. Convert server-side buffer population from a loop into a list comprehension to be more Pythonic
    1. Remove redundant waiting of the client for `220 ready`. Instead, we use a guarantee that `asyncio.loop.sock_connect` yields only after establishing a connection
    1. Remove `SendfileTestServer` with related asyncio imports
    1. Remove threading leftovers (`asyncio.to_thread` maintains its own threading machinery)

<!-- issue-number: [bpo-47015](https://bugs.python.org/issue47015) -->
https://bugs.python.org/issue47015
<!-- /issue-number -->

arhadthedev committed:
Remove an old, asyncore server

arhadthedev committed:
Remove threading infrastructure

arhadthedev committed:
Add a NEWS entry

arhadthedev committed:
Convert TestSendfile to IsolatedAsyncioTestCase

arhadthedev committed:
Remove redundant synchronization

arhadthedev committed:
Use an asyncio-connected client

arhadthedev committed:
Make os.sendfile asynchronous

arhadthedev committed:
Add and use an asyncio server

arhadthedev committed:
Turn client data reading loop into a comprehension

arhadthedev committed:
Merge branch 'main' into asyncio-test-os

arhadthedev committed:
Add an explicit asyncio event loop policy

merwok commented on Lib/test/test_os.py:
```diff
@@ -3328,40 +3238,52 @@ class TestSendfile(unittest.TestCase):
+    async def asyncSetUp(self):
+        self.server_buffer = b''
+        cb = lambda reader, writer: self.handle_new_client(reader, writer)
+        self.server = await asyncio.start_server(cb, socket_helper.HOSTv4)
```
It the callback is a function with `reader, writer` params, why not pass `self.handle_new_client` directly without the intervening lambda?

arhadthedev replied:
Thank you, I didn't know that Python autocaptures `self` when obtaining a reference to a method so I've done it manually in the lambda first. I've thought that, like in C++ or JavaScript, just getting a method reference without calling it right there will require to specify a call target once more.

merwok replied:
It’s the beauty of descriptors!

arhadthedev committed:
Address a review of merwok

asvetlov commented on Lib/test/test_os.py:
```diff
@@ -70,6 +64,8 @@
 from test.support.os_helper import FakePath
 
 
+asyncio.set_event_loop_policy(asyncio.DefaultEventLoopPolicy())
```
Why is this line needed?

asvetlov commented on Lib/test/test_os.py:
```diff
@@ -70,6 +64,8 @@
 
 
+asyncio.set_event_loop_policy(asyncio.DefaultEventLoopPolicy())
+
```
Please add the following to prevent a warning about unexpected environment modification.
```
def tearDownModule():
    asyncio.set_event_loop_policy(None)
```

arhadthedev replied:
Thank you, changed.

asvetlov commented on Lib/test/test_os.py:
```diff
@@ -3328,40 +3238,52 @@ class TestSendfile(unittest.TestCase):
-        self.client.recv(1024)
+        self.client.setblocking(False)
+        l = asyncio.get_running_loop()
+        await l.sock_connect(self.client, self.server.sockets[0].getsockname())
```
Don't hesitate to name a loop as the loop :)
```suggestion
        loop = asyncio.get_running_loop()
        await loop.sock_connect(self.client, self.server.sockets[0].getsockname())
```

arhadthedev replied:
The short name allowed to keep a 79 character limit. Though, on second thoughts, I took out a server address instead:

```diff
-        l = asyncio.get_running_loop()
-        await l.sock_connect(self.client, self.server.sockets[0].getsockname())
+        server_name = self.server.sockets[0].getsockname()
+        await asyncio.get_running_loop().sock_connect(self.client, server_name) 
```

arhadthedev replied:
On *the third thoughts*, I moved `server_name = self.server.sockets[0].getsockname()` right under server creation. Without it, reading code as a prose becomes harder:

> Start the server. For a client, create a socket, set it to be nonblocking, get a server name that is a name of its listening port, connect to the server.

instead of:

> Start the server, its name is a name of its listening port. For a client, create a socket, set it to be nonblocking, connect to the server.

arhadthedev committed:
Address a review of asvetlov

arhadthedev committed:
Group server stuff together

Without the grouping, reading code as a prose becomes harder:

> Start the server. For a client, create a socket, set it nonblocking,
> get a server name that is a name of its listening port, connect to the
> server.

instead of:

> Start the server, its name is a name of its listening port. For
> a client, create a socket, set it nonblocking, connect to the server.

merwok commented on Lib/test/test_os.py:
```diff
@@ -101,6 +95,10 @@ def create_file(filename, content=b'content'):
 
 
+def tearDownModule():
+    asyncio.set_event_loop_policy(None)
```
The line that set a loop policy is not present any more, so this cleanup step doesn’t seem to apply anymore.

arhadthedev replied:
If I understand [a failed pre-policy run](https://github.com/python/cpython/runs/5604663664?check_suite_focus=true) correctly, the policy must either be manually set to something before any case is runned or be returned back to `None` before the suit exits. I've chosen the first variant, Andrew proposed to switch to the second.

```plain
0:04:29 load avg: 3.56 [300/433/1] test_os failed (env changed) -- running: test_asyncio (1 min 10 sec), test_concurrent_futures (1 min 57 sec)

[a bunch of oks and skips follow]

[...]

423 tests OK.

10 slowest tests:
- test_tools: 4 min 44 sec
- test_concurrent_futures: 3 min 30 sec
- test_peg_generator: 2 min 57 sec
- test_multiprocessing_spawn: 2 min 27 sec
- test_gdb: 1 min 51 sec
- test_asyncio: 1 min 50 sec
- test_multiprocessing_forkserver: 1 min 38 sec
- test_multiprocessing_fork: 1 min 22 sec
- test_regrtest: 1 min 12 sec
- test_statistics: 49.1 sec

1 test altered the execution environment:
    test_os

9 tests skipped:
    test_devpoll test_ioctl test_kqueue test_msilib test_startfile
    test_winconsoleio test_winreg test_winsound test_zipfile64

Total duration: 11 min 50 sec
Tests result: ENV CHANGED
make: *** [Makefile:1795: buildbottest] Error 3
Error: Process completed with exit code 2.
```

arhadthedev committed:
Fix a call of `async_sendfile` in `test_flags`
